python-package/lightgbm/callback.py

```diff
@@ -209,12 +215,69 @@ def _init(env):
+                        ("xentropy", {"cross_entropy"}),
+                        ("xentlambda", {"cross_entropy_lambda"}),
+                        ("kldiv", {"kullback_leibler"})]
+        for metric, aliases in metrics_list:
+            if target_metric in aliases:
+                return metric
+        return target_metric
+
     def _callback(env):
         if not cmp_op:
             _init(env)
         if not enabled[0]:
             return
+        if first_metric_only:
+            # Choosing a target metric for early stopping from first element of metrics list as `eval_metric`
+            # if first_metric_only == True
+            eval_metric = None
+            for metric_alias in ['metric', 'metrics', 'metric_types']:
+                if metric_alias in env.params.keys():
+                    if isinstance(env.params[metric_alias], (tuple, list)):
+                        metric_list = [m for m in env.params[metric_alias] if m is not None]
+                        eval_metric = metric_list[0]
+                    else:  # string
+                        if env.evaluation_result_list[0][0] == "cv_agg":  # for lgb.cv
+                            for i in range(len(env.evaluation_result_list)):
+                                # Removing training data
+                                if ((env.evaluation_result_list[i][1].find(" ") >= 0
+                                     and env.evaluation_result_list[i][1].split(" ")[0] != "train")):
+                                    eval_metric = env.evaluation_result_list[i][1].split(" ")[1]
+                                    break
+                        else:  # the other than lgb.cv
+                            eval_metric = env.params[metric_alias]
+                    break
+            if eval_metric is None:
+                # If `eval_metric` was not chosen from `env.params`,
+                # using first element of env.evaluation_result_list for early stopping
+                eval_metric = env.evaluation_result_list[0][1]
+            eval_metric = _metric_alias_matching(eval_metric)
         for i in range_(len(env.evaluation_result_list)):
+            metric_key = env.evaluation_result_list[i][1]
+            if env.evaluation_result_list[i][0] == "cv_agg":  # for lgb.cv
+                if metric_key.split(" ")[0] == "train":
+                    continue  # train metric is not used on early stopping
+                if ((first_metric_only and eval_metric is not None and eval_metric != ""
+                     and metric_key != "valid {}".format(eval_metric) and metric_key != eval_metric)):
+                    continue
+            elif env.evaluation_result_list[i][0] == "training":
+                continue  # for training set, early stopping is not applied
+            else:  # for lgb.train and sklearn wrapper
+                if first_metric_only:
+                    if metric_key != eval_metric:
```

Comment:
@matsuken92 I haven't checked yet, but is everything OK with metrics with composite names in results, e.g. ``ndcg@1`` or ``multi_error@3``?

https://github.com/microsoft/LightGBM/blob/16751b3803e847592dba3bab876c4f052507e45d/src/metric/rank_metric.hpp#L40-L42
https://github.com/microsoft/LightGBM/blob/16751b3803e847592dba3bab876c4f052507e45d/src/metric/multiclass_metric.hpp#L151-L154